fix: UriBuilder queryParam and replaceQueryParam should ignore null values

Crow-EH opened the pull request:
Fix `UriBuilder::queryParam` and `UriBuilder::replaceQueryParam` to hold its promise on null values.

According to the javadoc of [`UriBuilder::queryParam`](https://github.com/micronaut-projects/micronaut-core/blob/3.6.x/http/src/main/java/io/micronaut/http/uri/UriBuilder.java#L89) and `UriBuilder::replaceQueryParam`:
> If either name or values are null the value will be ignored.

In Java (at least), passing `null` to `UriBuilder::queryParam` among other valid query parameters will produce dangling `&` in the final URI. This is because when calling `builder.queryParam('foo', null)`, the resulting parameter `Object... values` will be an array of size 1 containing `null`.

Also fixes the case where values is an array of mutliple `null`.

Example:
```
UriBuilder.of('/foo')
    .queryParam("foo", "ok")
    .queryParam("bar", null)
    .queryParam("bar", anArrayOfNull)
    .queryParam("fizz", "fine")
    .build()
    .toString();
```

Expected result: `/foo?foo=ok&fizz=fine`

Actual result: `/foo?foo=ok&&&fizz=fine`

I only tested the `[null, null]` case since it seems Groovy doesn't have the same behaviour than Java when passing only`null`: `Object... values` is `null` and correctly ignored.

Testing the whole issue would require a Java test, I can add one if you want.

Crow-EH committed:
fix: UriBuilder queryParam and replaceQueryParam should ignore null values

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-core?pullRequest=7681) <br/>All committers have signed the CLA.

Crow-EH commented:
Here's an example of what I have to do to work around this issue :
```
if (!result.next().isEmpty()) {
  var uriBuilder = UriBuilder.of(host.replace("http:", "https:") + uri)
      .queryParam("order", parameters.order())
      .queryParam("size", parameters.size())
      .queryParam("searchAfter", result.next().toArray());
  // TODO(ysiguret): stop checking for null when
  // https://github.com/micronaut-projects/micronaut-core/pull/7681 is released
  if (isNotEmpty(parameters.states())) {
    uriBuilder.queryParam("states", parameters.states().toArray());
  }
  if (parameters.minPrice() != null) {
    uriBuilder.queryParam("minPrice", parameters.minPrice());
  }
  if (parameters.maxPrice() != null) {
    uriBuilder.queryParam("maxPrice", parameters.maxPrice());
  }
  if (parameters.ownerId() != null) {
    uriBuilder.queryParam("ownerId", parameters.ownerId());
  }
  if (parameters.externalUserId() != null) {
    uriBuilder.queryParam("externalUserId", parameters.externalUserId());
  }
  if (parameters.sortByAttribute() != null) {
    uriBuilder.queryParam("sortByAttribute", parameters.sortByAttribute());
  }
  if (parameters.sortAttributeType() != null) {
    uriBuilder.queryParam("sortAttributeType", parameters.sortAttributeType());
  }
  if (parameters.sortBy() != null) {
    uriBuilder.queryParam("sortBy", parameters.sortBy().path());
  }

  responseBuilder.next(Link.builder().rel(Rel.Next).href(uriBuilder.build()).build());
}
```

event though I should be able to do that according to the documentation :
```
if (!result.next().isEmpty()) {
  var uriBuilder = UriBuilder.of(host.replace("http:", "https:") + uri)
      .queryParam("order", parameters.order())
      .queryParam("size", parameters.size())
      .queryParam("searchAfter", result.next().toArray())
      .queryParam("states", isNotEmpty(parameters.states()) ? parameters.states().toArray() : null)
      .queryParam("minPrice", parameters.minPrice())
      .queryParam("maxPrice", parameters.maxPrice())
      .queryParam("ownerId", parameters.ownerId())
      .queryParam("externalUserId", parameters.externalUserId())
      .queryParam("sortByAttribute", parameters.sortByAttribute())
      .queryParam("sortAttributeType", parameters.sortAttributeType())
      .queryParam("sortBy", parameters.sortBy() != null ? parameters.sortBy().path() : null);

  responseBuilder.next(Link.builder().rel(Rel.Next).href(uriBuilder.build()).build());
}
```

Crow-EH commented:
Added the same fix to `UriBuilder::replaceQueryParam` since it had the same issue.

Also added a test since the method was not tested.

Although I don't know if I should continue rebasing my branch regularly, or if I'm missing something and no one will ever look at this PR. 😅

alvarosanchez commented:
@Crow-EH sorry for taking so long to address your PR, and thank you very much for your contribution!
